### ISSUE TITLE
parameter setting problem in keras?

### ISSUE THREAD

**Issue by BIGBALLON** [uber/horovod#114]
assume I have an CNN network trained by **single machine(1 gpu):** 
> epoch : 180
batch size:64
iteration: 200

then I want to use mult-gpu (4 gpus, 1 gpu each machine) to train, 
**only** need to achieve **the same** performance.

which one is **correct**!

**1.**  batch size divided 4
> epoch : 180
batch size:16
iteration: 200

**2**  iteration divided 4
> epoch : 180
batch size:64
iteration: 50

**3** epoch divided 4
> epoch : 45
batch size:64
iteration: 200

I thinks **1**, **2**, **3** are all correct
but **1** is reasonable.
(sometimes 1GPU can't support large batch size, that is why we need to use multi-gpu or distributed gpus )

**but it seems that  there are some unfriendly logs in keras ?**
see the iterations: **1563/391** 
```bash
1563/391 [=======================================================================================================================] - 19s 12ms/step - loss: 1.7795 - acc: 0.3749 - val_loss: 1.6854 - val_acc: 0.4244oss: 1.8112 - acc: 0.3494
Epoch 3/200
1563/391 [=======================================================================================================================] - 19s 12ms/step - loss: 1.7868 - acc: 0.3765 - val_loss: 1.6845 - val_acc: 0.4152
Epoch 3/200
1563/391 [=======================================================================================================================] - 19s 12ms/step - loss: 1.7858 - acc: 0.3790 - val_loss: 1.6707 - val_acc: 0.4328
Epoch 3/200
1563/391 [=======================================================================================================================] - 19s 12ms/step - loss: 1.7952 - acc: 0.3665 - val_loss: 1.6924 - val_acc: 0.4207
Epoch 3/200
128/391 [========>.....................] - ETA: 3s - loss: 1.7661 - acc: 0.4038128/391 [========>.....................] - ETA: 3s - loss: 1.7695 - acc: 0.3857128/391 [========>.....................] - ETA: 3s - loss: 1.7199 - acc: 0.406712387/391 [============================>.] - ETA: 0s - loss: 1.7772 - acc: 0.3840387/391 [============================>.] - ETA: 0s - lo
```

```
 4/117 [>.............................] - ETA: 11s - loss: 0.1845 - acc: 0.9355  4/117 [>.............................] - ETA: 11s - loss: 0.1079 - acc: 0.9668  4/117 [>.............................] - ETA: 11s - loss: 0.1292 - acc: 0.951  7/117 [>.............................] - ETA: 9s - loss: 0.2120 - acc: 0.9330   7/117 [>.............................] - ETA: 9s - loss: 0.1600 - acc: 0.9509   7/117 [>.............................] - ETA: 8s - loss: 0.1360 - acc: 0.9565 23/117 [====>.........................] - ETA: 7s - loss: 0.1576 - acc: 0.9518 23/117 [====>.........................] - ETA: 7s - loss: 0.1581 - acc: 0.9487 23/117 [====>.........................] - ETA: 7s - loss: 0.1683 - acc: 0.9463 2 32/117 [=======>......................] - ETA: 6s - loss: 0.1561 - acc: 0.9521 32/117 [=======>......................] - ETA: 6s - loss: 0.1603 - acc: 0.9500 32/117 [=======>......................] - ETA: 6s - loss: 0.1647 - acc: 0.9490 3 34/117 [=======>......................] - ETA: 6s - loss: 0.1585 - acc: 0.9517 34/117 [=======>......................] - ETA: 6s - loss: 0.1613 - acc: 0.9499 34/117 [=======>......................] - ETA: 6s - loss: 0.1671 - acc: 0.9483 3 47/117 [===========>..................] - ETA: 5s - loss: 0.1553 - acc: 0.9513 47/117 [===========>..................] - ETA: 5s - loss: 0.1658 - acc: 0.9485 47/117 [===========>..................] - ETA: 5s - loss: 0.1646 - acc: 0.9498 4111/117 [===========================>..] - ETA: 0s - loss: 0.1490 - acc: 0.9552111/117 [===========================>..] - ETA: 0s - loss: 0.1623 - acc: 0.9512111/117 [===========================>..] - ETA: 0s - loss: 0.1599 - acc: 0.951611469/117 [========================================================================================================================] - 35s 76ms/step - loss: 0.1466 - acc: 0.9565 - val_loss: 0.0308 - val_acc: 0.9899oss: 0.1605 - acc: 0.9520
Epoch 3/24
469/117 [========================================================================================================================] - 35s 76ms/step - loss: 0.1448 - acc: 0.9575 - val_loss: 0.0308 - val_acc: 0.9899
Epoch 3/24
469/117 [========================================================================================================================] - 35s 76ms/step - loss: 0.1450 - acc: 0.9567 - val_loss: 0.0308 - val_acc: 0.9899
Epoch 3/24
469/117 [========================================================================================================================] - 35s 76ms/step - loss: 0.1426 - acc: 0.9570 - val_loss: 0.0308 - val_acc: 0.9899
Epoch 3/24
  5/117 [>.............................] - ETA: 9s - loss: 0.1439 - acc: 0.9516   5/117 [>.............................] - ETA: 9s - loss: 0.1218 - acc: 0.9516   5/117 [>.............................] - ETA: 9s - loss: 0.0867 - acc: 0.9719  6/117 [>.............................] - ETA: 9s - loss: 0.1240 - acc: 0.9518  6/117 [>.............................] - ETA: 8s - loss: 0.1110 - acc: 0.9674  6/117 [>.............................] - ETA: 9s - loss: 0.1513 - acc: 0.9492   37/117 [========>.....................] - ETA: 6s - loss: 0.1307 - acc: 0.9628 37/117 [========>.....................] - ETA: 5s - loss: 0.1293 - acc: 0.9633 37/117 [========>.....................] - ETA: 6s - loss: 0.1188 - acc: 0.9628 3 40/117 [=========>....................] - ETA: 5s - loss: 0.1284 - acc: 0.9633 40/117 [=========>....................] - ETA: 5s - loss: 0.1167 - acc: 0.9613 40/117 [=========>....................] - ETA: 5s - loss: 0.1299 - acc: 0.9633 4 55/117 [=============>................] - ETA: 4s - loss: 0.1261 - acc: 0.9634 55/117 [=============>................] - ETA: 4s - loss: 0.1210 - acc: 0.9632 55/117 [=============>................] - ETA: 4s - loss: 0.1248 - acc: 0.9612 5469/117 [========================================================================================================================] - 35s 75ms/step - loss: 0.1186 - acc: 0.9640 - val_loss: 0.0222 - val_acc: 0.9929oss: 0.1171 - acc: 0.9643
Epoch 4/24
469/117 [========================================================================================================================] - 35s 75ms/step - loss: 0.1177 - acc: 0.9653 - val_loss: 0.0222 - val_acc: 0.9929
Epoch 4/24
469/117 [========================================================================================================================] - 35s 75ms/step - loss: 0.1205 - acc: 0.9641 - val_loss: 0.0222 - val_acc: 0.9929
Epoch 4/24
469/117 [========================================================================================================================] - 35s 75ms/step - loss: 0.1174 - acc: 0.9649 - val_loss: 0.0222 - val_acc: 0.9929
Epoch 4/24
 71/117 [=================>............] - ETA: 3s - loss: 0.1153 - 
```

**Comment by alsrgv** [comment]
@BIGBALLON, I'm glad we're past setup issue and now are chatting about math behind distributed deep learning :-)

In practice, approach 1 does not scale since GPUs are not optimized to train small batches.  This will also make communication overhead much larger, since you still need to exchange the whole model, regardless of what is your batch size.

So, most people use approach 2.  In case of approach 2, your logical batch size would be single_gpu_batch_size * number_of_gpus.  In order to make model converge faster due to this larger batch, you need to adjust learning rate.  Most popular way to do so is to just multiple LR by the number of GPUs.

If you have a lot of GPUs, starting with super high learning rate that resulted from multiplication above may harm your final accuracy. So, Facebook [recommended](https://research.fb.com/publications/accurate-large-minibatch-sgd-training-imagenet-in-1-hour/) doing a gradual warmup of learning rate during first few epochs.

[MNIST advanced example](https://github.com/uber/horovod/blob/master/examples/keras_mnist_advanced.py) covers this learning rate multiplication, warmup, and adjustment of # of iterations.

**Comment by BIGBALLON** [comment]
@alsrgv thank you， alsrgv。

**single machine(1 gpu):**
>epoch : 180
batch size:64
iteration: 200

I tried approach  2, it seems that we just use [45 epochs, batch size 64, iteration 50] can get the same performance as single machine(1 gpu).

I am confused why is 45 epochs, not 185 epochs ?
I only change the iteration from 200 to 50.

**Comment by BIGBALLON** [comment]
@alsrgv  this is my tony demo code:

```python
import keras
import numpy as np
from keras import optimizers
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Conv2D, Dense, Flatten, MaxPooling2D
from keras.callbacks import LearningRateScheduler, TensorBoard
from keras.preprocessing.image import ImageDataGenerator
from keras.regularizers import l2
from keras import backend as K
import math
import tensorflow as tf
import horovod.keras as hvd

# Initialize Horovod.
hvd.init()

# Pin GPU to be used to process local rank (one GPU per process)
config = tf.ConfigProto()
config.gpu_options.allow_growth = True
config.gpu_options.visible_device_list = str(hvd.local_rank())
K.set_session(tf.Session(config=config))


batch_size    = 128
epochs        = 200
iterations    = 391
num_classes   = 10
weight_decay  = 0.0001
mean          = [125.307, 122.95, 113.865]
std           = [62.9932, 62.0887, 66.7048]

def build_model():
    model = Sequential()
    model.add(Conv2D(6, (5, 5), padding='valid', activation = 'relu', kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay), input_shape=(32,32,3)))
    model.add(MaxPooling2D((2, 2), strides=(2, 2)))
    model.add(Conv2D(16, (5, 5), padding='valid', activation = 'relu', kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay)))
    model.add(MaxPooling2D((2, 2), strides=(2, 2)))
    model.add(Flatten())
    model.add(Dense(120, activation = 'relu', kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay) ))
    model.add(Dense(84, activation = 'relu', kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay) ))
    model.add(Dense(10, activation = 'softmax', kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay) ))
    return model

def scheduler(epoch):
    if epoch <= 60:
        return 0.08
    if epoch <= 120:
        return 0.01
    if epoch <= 160:    
        return 0.002
    return 0.0004

if __name__ == '__main__':

    # load data
    (x_train, y_train), (x_test, y_test) = cifar10.load_data()
    y_train = keras.utils.to_categorical(y_train, num_classes)
    y_test = keras.utils.to_categorical(y_test, num_classes)
    x_train = x_train.astype('float32')
    x_test = x_test.astype('float32')
    
    # data preprocessing  [raw - mean / std]
    for i in range(3):
        x_train[:,:,:,i] = (x_train[:,:,:,i] - mean[i]) / std[i]
        x_test[:,:,:,i] = (x_test[:,:,:,i] - mean[i]) / std[i]

    model = build_model()
    sgd = optimizers.SGD(lr=.1, momentum=0.9, nesterov=True)
    sgd = hvd.DistributedOptimizer(sgd)

    model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])

    # set callback
    cbks = [
            hvd.callbacks.BroadcastGlobalVariablesCallback(0),
            LearningRateScheduler(scheduler)]


    # using real-time data augmentation
    print('Using real-time data augmentation.')
    datagen = ImageDataGenerator(horizontal_flip=True,
            width_shift_range=0.125,height_shift_range=0.125,fill_mode='constant',cval=0.)

    datagen.fit(x_train)

    # start traing 
    model.fit_generator(datagen.flow(x_train, y_train,batch_size=batch_size ),
                        steps_per_epoch=iterations // hvd.size(),
                        epochs=epochs,
                        callbacks=cbks,
                        validation_data=(x_test, y_test))
```

Is there something wrong with me?

**Comment by alsrgv** [comment]
@BIGBALLON, can you try downgrading your Keras to 2.0.8?  2.1.1 has an issue (https://github.com/fchollet/keras/pull/8509) that it ignores `steps_per_epoch`.

I played with your learning rate schedule a bit and got a tad better accuracy on 4 GPUs, in 1/4th time (3s per epoch instead of 12s per epoch on single GPU):
```
Epoch 200/200
97/97 [==============================] - 3s - loss: 0.7356 - acc: 0.7609 - val_loss: 0.7557 - val_acc: 0.7580
```

I modified your `scheduler` and `lr` of `optimizers.SGD` to be more conservative.  LR you were using seemed a bit too high.  At the same time, I added multiplication or LR by the `hvd.size()`, which is common practice.  I also added warmup callback, so during first 30 epochs LR gradually rises from 0.01 to `0.01 * hvd.size()`.

Here's full code:
```python
import keras
import numpy as np
from keras import optimizers
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers import Conv2D, Dense, Flatten, MaxPooling2D
from keras.callbacks import LearningRateScheduler, TensorBoard
from keras.preprocessing.image import ImageDataGenerator
from keras.regularizers import l2
from keras import backend as K
import math
import tensorflow as tf
import horovod.keras as hvd

# Initialize Horovod.
hvd.init()

# Pin GPU to be used to process local rank (one GPU per process)
config = tf.ConfigProto()
config.gpu_options.allow_growth = True
config.gpu_options.visible_device_list = str(hvd.local_rank())
K.set_session(tf.Session(config=config))


batch_size    = 128
epochs        = 200
iterations    = 391
num_classes   = 10
weight_decay  = 0.0001
mean          = [125.307, 122.95, 113.865]
std           = [62.9932, 62.0887, 66.7048]

def build_model():
    model = Sequential()
    model.add(Conv2D(6, (5, 5), padding='valid', activation = 'relu', kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay), input_shape=(32,32,3)))
    model.add(MaxPooling2D((2, 2), strides=(2, 2)))
    model.add(Conv2D(16, (5, 5), padding='valid', activation = 'relu', kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay)))
    model.add(MaxPooling2D((2, 2), strides=(2, 2)))
    model.add(Flatten())
    model.add(Dense(120, activation = 'relu', kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay) ))
    model.add(Dense(84, activation = 'relu', kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay) ))
    model.add(Dense(10, activation = 'softmax', kernel_initializer='he_normal', kernel_regularizer=l2(weight_decay) ))
    return model

def scheduler(epoch):
    if epoch <= 30:
        # bypass to the warmup callback
        return K.get_value(model.optimizer.lr)
    if epoch <= 60:
        return 0.01 * hvd.size()
    if epoch <= 120:
        return 0.002 * hvd.size()
    if epoch <= 160:
        return 0.0004 * hvd.size()
    return 0.00008 * hvd.size()

if __name__ == '__main__':

    # load data
    (x_train, y_train), (x_test, y_test) = cifar10.load_data()
    y_train = keras.utils.to_categorical(y_train, num_classes)
    y_test = keras.utils.to_categorical(y_test, num_classes)
    x_train = x_train.astype('float32')
    x_test = x_test.astype('float32')

    # data preprocessing  [raw - mean / std]
    for i in range(3):
        x_train[:,:,:,i] = (x_train[:,:,:,i] - mean[i]) / std[i]
        x_test[:,:,:,i] = (x_test[:,:,:,i] - mean[i]) / std[i]

    model = build_model()
    sgd = optimizers.SGD(lr=.01 * hvd.size(), momentum=0.9, nesterov=True)
    sgd = hvd.DistributedOptimizer(sgd)

    model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])

    # set callback
    cbks = [
            hvd.callbacks.BroadcastGlobalVariablesCallback(0),
            hvd.callbacks.LearningRateWarmupCallback(warmup_epochs=30, verbose=1),
            LearningRateScheduler(scheduler)]

    # using real-time data augmentation
    print('Using real-time data augmentation.')
    datagen = ImageDataGenerator(horizontal_flip=True,
            width_shift_range=0.125,height_shift_range=0.125,fill_mode='constant',cval=0.)

    datagen.fit(x_train)

    # start traing
    model.fit_generator(datagen.flow(x_train, y_train,batch_size=batch_size ),
                        steps_per_epoch=iterations // hvd.size(),
                        epochs=epochs,
                        callbacks=cbks,
                        validation_data=(x_test, y_test))
```

**Comment by alsrgv** [comment]
@BIGBALLON, you can now also upgrade to Keras 2.1.2 which was released today and has that bug fixed :-)

**Comment by BIGBALLON** [comment]
@alsrgv 
thanks a lot!!

I upgraded keras to the 2.1.2, then my demo works perfectly.
Sincere thanks for your help and guidance, you solved all my questions!
awesome!

**Comment by xKelsius** [comment]
@alsrgv Related to this topic, if you adjust the number of iterations as seen in approach 2 above, and you're using a sorted, ordered dataset, would this essentially divide the dataset by the number of GPUs? Such as if you have 1000 images in your training dataset for 4 GPUs, would each process/GPU only process the 1st 250 images in your data generator?

If this is the case, would you recommend randomizing the order of your data when creating your data generator?

**Comment by alsrgv** [comment]
@kelsiuss, with Keras, I would say the easiest option is to randomly read the dataset on all the workers, like we do in ImageNet example - https://github.com/uber/horovod/blob/master/examples/keras_imagenet_resnet50.py.

The option of pre-partitioning and re-shuffling dataset on every epoch is feasible but requires more work.  I've tried both options and arrived to the same final accuracy, so the recommendation is to use whatever is easiest.
